File: client/go/ReverseWebSocketUpstreamTransport.go
package modcdp

import (
	"encoding/json"
	"fmt"
	"net"
	"net/url"
	"sync"
	"time"

	"github.com/gobwas/ws"
	"github.com/gobwas/ws/wsutil"
)

const DefaultReverseWSBind = "127.0.0.1:29292"
const DefaultReverseWSWaitTimeoutMS = 10_000

type ReverseWebSocketUpstreamTransport struct {
	UpstreamTransport
	Bind          string
	URL           string
	WaitTimeoutMS int
	Listener      net.Listener
	Conn          net.Conn
	writeMu       sync.Mutex
	peerCh        chan struct{}
	peerOnce      sync.Once
	PeerInfo      map[string]any
}

func NewReverseWebSocketUpstreamTransport(bind string, wait_timeout_ms int) *ReverseWebSocketUpstreamTransport {
	if bind == "" {
		bind = DefaultReverseWSBind
	}
	if wait_timeout_ms == 0 {
		wait_timeout_ms = DefaultReverseWSWaitTimeoutMS
	}
	t := &ReverseWebSocketUpstreamTransport{WaitTimeoutMS: wait_timeout_ms, peerCh: make(chan struct{})}
	t.SetBind(bind)
	return t
}

func (t *ReverseWebSocketUpstreamTransport) SetBind(bind string) {
	parsed, err := url.Parse(bind)
	if err != nil || parsed.Scheme == "" {
		parsed, _ = url.Parse("ws://" + bind)
	}
	host := parsed.Hostname()
	if host == "" {
		host = "127.0.0.1"
	}
	port := parsed.Port()
	if port == "" {
		port = "29292"
	}
	t.Bind = net.JoinHostPort(host, port)
	t.URL = "ws://" + t.Bind
}

func (t *ReverseWebSocketUpstreamTransport) Update(config map[string]any) {
	if config == nil {
		return
	}
	if bind, _ := config["reversews_bind"].(string); bind != "" {
		t.SetBind(bind)
	} else if rawURL, _ := config["url"].(string); rawURL != "" {
		t.SetBind(rawURL)
	}
	if waitTimeoutMS, ok := intFromConfig(config["reversews_wait_timeout_ms"]); ok {
		t.WaitTimeoutMS = waitTimeoutMS
	}
}

func (t *ReverseWebSocketUpstreamTransport) Connect() error {
	listener, err := net.Listen("tcp", t.Bind)
	if err != nil {
		return err
	}
	t.Listener = listener
	go t.acceptLoop()
	return nil
}

func (t *ReverseWebSocketUpstreamTransport) Send(message map[string]any) error {
	if t.Conn == nil {
		return fmt.Errorf("no reverse ModCDP extension peer is connected at %s", t.URL)
	}
	body, err := json.Marshal(message)
	if err != nil {
		return err
	}
	t.writeMu.Lock()
	defer t.writeMu.Unlock()
	return wsutil.WriteServerText(t.Conn, body)
}

func (t *ReverseWebSocketUpstreamTransport) GetInjectorConfig() ExtensionInjectorConfig {
	return ExtensionInjectorConfig{ReverseProxyURL: t.URL}
}

func (t *ReverseWebSocketUpstreamTransport) WaitForPeer() error {
	if t.Conn != nil {
		return nil
	}
	select {
	case <-t.peerCh:
		return nil
	case <-time.After(time.Duration(t.WaitTimeoutMS) * time.Millisecond):
		return fmt.Errorf("timed out waiting %dms for reverse ModCDP extension connection", t.WaitTimeoutMS)
	}
}

func (t *ReverseWebSocketUpstreamTransport) Close() error {
	if t.Conn != nil {
		_ = t.Conn.Close()
		t.Conn = nil
	}
	if t.Listener != nil {
		_ = t.Listener.Close()
		t.Listener = nil
	}
	return nil
}

func (t *ReverseWebSocketUpstreamTransport) acceptLoop() {
	for {
		conn, err := t.Listener.Accept()
		if err != nil {
			return
		}
		go t.accept(conn)
	}
}

func (t *ReverseWebSocketUpstreamTransport) accept(conn net.Conn) {
	if _, err := ws.Upgrade(conn); err != nil {
		_ = conn.Close()
		t.emitClose(err)
		return
	}
	_ = conn.SetReadDeadline(time.Now().Add(time.Duration(t.WaitTimeoutMS) * time.Millisecond))
	helloBytes, err := wsutil.ReadClientText(conn)
	_ = conn.SetReadDeadline(time.Time{})
	if err != nil {
		_ = conn.Close()
		t.emitClose(err)
		return
	}
	var hello map[string]any
	if err := json.Unmarshal(helloBytes, &hello); err != nil || hello["type"] != "modcdp.reverse.hello" {
		_ = conn.Close()
		if err == nil {
			err = fmt.Errorf("invalid reverse hello")
		}
		t.emitClose(err)
		return
	}
	if t.Conn != nil && t.Conn != conn {
		_ = t.Conn.Close()
	}
	t.Conn = conn
	t.PeerInfo = hello
	t.peerOnce.Do(func() { close(t.peerCh) })
	for {
		data, err := wsutil.ReadClientText(conn)
		if err != nil {
			if t.Conn == conn {
				t.Conn = nil
				t.PeerInfo = nil
			}
			t.emitClose(err)
			return
		}
		var message map[string]any
		if err := json.Unmarshal(data, &message); err == nil {
			t.emitRecv(message)
		}
	}
}

File: client/python/modcdp/ReverseWebSocketUpstreamTransport.py
from __future__ import annotations

import base64
import hashlib
import json
import socket
import struct
import threading
from typing import Any
from urllib.parse import urlparse

from .UpstreamTransport import UpstreamTransport


DEFAULT_REVERSEWS_BIND = "127.0.0.1:29292"
DEFAULT_REVERSEWS_WAIT_TIMEOUT_MS = 10_000
_WS_GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"


class ReverseWebSocketUpstreamTransport(UpstreamTransport):
    mode = "reversews"
    endpoint_kind = "modcdp_server"

    def __init__(self, bind: str = DEFAULT_REVERSEWS_BIND, wait_timeout_ms: int = DEFAULT_REVERSEWS_WAIT_TIMEOUT_MS) -> None:
        super().__init__()
        self.wait_timeout_ms = wait_timeout_ms
        self.server_socket: socket.socket | None = None
        self.socket: socket.socket | None = None
        self.accept_thread: threading.Thread | None = None
        self.reader_thread: threading.Thread | None = None
        self.peer_info: dict[str, Any] | None = None
        self.peer_event = threading.Event()
        self.closed = False
        self.write_lock = threading.Lock()
        self.setBind(bind)

    def update(self, config: dict[str, Any] | None = None) -> "ReverseWebSocketUpstreamTransport":
        config = config or {}
        bind = config.get("reversews_bind") or config.get("url")
        if isinstance(bind, str) and bind:
            self.setBind(bind)
        return self

    def getInjectorConfig(self) -> dict[str, Any]:
        return {"reverse_proxy_url": self.url}

    def setBind(self, bind: str) -> None:
        parsed = urlparse(bind if "://" in bind else f"ws://{bind}")
        host = parsed.hostname or "127.0.0.1"
        port = parsed.port or 29292
        if port <= 0 or port > 65535:
            raise ValueError(f"Invalid host:port {bind}")
        self.url = f"ws://{host}:{port}"

    def connect(self) -> None:
        parsed = urlparse(self.url or "")
        host = parsed.hostname or "127.0.0.1"
        port = parsed.port or 29292
        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server_socket.bind((host, port))
        server_socket.listen(1)
        self.server_socket = server_socket
        self.closed = False
        self.accept_thread = threading.Thread(target=self._accept_loop, daemon=True)
        self.accept_thread.start()

    def send(self, message: dict[str, Any]) -> None:
        if self.socket is None:
            raise RuntimeError(f"No reverse ModCDP extension peer is connected at {self.url}.")
        payload = json.dumps(message).encode()
        with self.write_lock:
            self.socket.sendall(_encode_server_text_frame(payload))

    def wait_for_peer(self) -> None:
        if self.socket is not None:
            return
        if not self.peer_event.wait(self.wait_timeout_ms / 1000):
            raise RuntimeError(f"Timed out waiting {self.wait_timeout_ms}ms for reverse ModCDP extension connection.")

    def close(self) -> None:
        self.closed = True
        for sock in (self.socket, self.server_socket):
            try:
                sock.close() if sock is not None else None
            except Exception:
                pass
        self.socket = None
        self.server_socket = None
        self.peer_info = None
        self.peer_event.clear()

    def _accept_loop(self) -> None:
        while not self.closed and self.server_socket is not None:
            try:
                sock, _ = self.server_socket.accept()
                self._accept(sock)
            except OSError:
                return
            except Exception as error:
                self._emit_close(error if isinstance(error, Exception) else RuntimeError(str(error)))

    def _accept(self, sock: socket.socket) -> None:
        try:
            _perform_server_handshake(sock)
            hello_raw = _read_client_text_frame(sock, self.wait_timeout_ms / 1000)
            if hello_raw is None:
                raise RuntimeError("reverse hello socket closed")
            hello = json.loads(hello_raw)
            if not isinstance(hello, dict) or hello.get("type") != "modcdp.reverse.hello":
                raise RuntimeError("invalid reverse hello")
            old_socket = self.socket
            if old_socket is not None and old_socket is not sock:
                try:
                    old_socket.close()
                except Exception:
                    pass
            self.socket = sock
            self.peer_info = hello
            self.peer_event.set()
            self.reader_thread = threading.Thread(target=self._read_loop, args=(sock,), daemon=True)
            self.reader_thread.start()
        except Exception as error:
            try:
                sock.close()
            except Exception:
                pass
            self._emit_close(error if isinstance(error, Exception) else RuntimeError(str(error)))

    def _read_loop(self, sock: socket.socket) -> None:
        try:
            while not self.closed and self.socket is sock:
                data = _read_client_text_frame(sock, None)
                if data is None:
                    break
                self._parse_and_emit_recv(data)
        except Exception as error:
            if not self.closed:
                self._emit_close(error if isinstance(error, Exception) else RuntimeError(str(error)))
        finally:
            if self.socket is sock:
                self.socket = None
                self.peer_info = None
                self.peer_event.clear()


def _perform_server_handshake(sock: socket.socket) -> None:
    request = b""
    while b"\r\n\r\n" not in request:
        chunk = sock.recv(4096)
        if not chunk:
            raise RuntimeError("websocket handshake closed")
        request += chunk
    headers: dict[str, str] = {}
    for line in request.decode(errors="replace").split("\r\n")[1:]:
        if ":" in line:
            key, value = line.split(":", 1)
            headers[key.strip().lower()] = value.strip()
    ws_key = headers.get("sec-websocket-key")
    if not ws_key:
        raise RuntimeError("websocket handshake missing Sec-WebSocket-Key")
    accept = base64.b64encode(hashlib.sha1((ws_key + _WS_GUID).encode()).digest()).decode()
    sock.sendall(
        (
            "HTTP/1.1 101 Switching Protocols\r\n"
            "Upgrade: websocket\r\n"
            "Connection: Upgrade\r\n"
            f"Sec-WebSocket-Accept: {accept}\r\n"
            "\r\n"
        ).encode()
    )


def _read_exact(sock: socket.socket, length: int) -> bytes:
    chunks: list[bytes] = []
    remaining = length
    while remaining > 0:
        chunk = sock.recv(remaining)
        if not chunk:
            raise EOFError("websocket closed")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)


def _read_client_text_frame(sock: socket.socket, timeout_s: float | None) -> str | None:
    old_timeout = sock.gettimeout()
    if timeout_s is not None:
        sock.settimeout(timeout_s)
    try:
        header = _read_exact(sock, 2)
        opcode = header[0] & 0x0F
        masked = bool(header[1] & 0x80)
        length = header[1] & 0x7F
        if length == 126:
            length = struct.unpack("!H", _read_exact(sock, 2))[0]
        elif length == 127:
            length = struct.unpack("!Q", _read_exact(sock, 8))[0]
        mask = _read_exact(sock, 4) if masked else b""
        payload = bytearray(_read_exact(sock, length))
        if masked:
            for index, value in enumerate(payload):
                payload[index] = value ^ mask[index % 4]
        if opcode == 0x8:
            return None
        if opcode != 0x1:
            return ""
        return payload.decode()
    finally:
        sock.settimeout(old_timeout)


def _encode_server_text_frame(payload: bytes) -> bytes:
    if len(payload) < 126:
        header = bytes([0x81, len(payload)])
    elif len(payload) < 65536:
        header = bytes([0x81, 126]) + struct.pack("!H", len(payload))
    else:
        header = bytes([0x81, 127]) + struct.pack("!Q", len(payload))
    return header + payload
